compiler/src/dotty/tools/dotc/config/PathResolver.scala

```diff
@@ -36,17 +36,25 @@ object PathResolver {
   /** Values found solely by inspecting environment or property variables.
    */
   object Environment {
-    private def searchForBootClasspath = (
-      systemProperties find (_._1 endsWith ".boot.class.path") map (_._2) getOrElse ""
-    )
+    private def searchForBootClasspath = {
+      import scala.jdk.CollectionConverters.*
+      val props = System.getProperties
+      // This formulation should be immune to ConcurrentModificationExceptions when system properties
+      // we're unlucky enough to witness a partially published result of System.setProperty or direct
+      // mutation of the System property map. stringPropertyNames internally uses the Enumeration interface,
+      // rather than Iterator, and this disables the fail-fast ConcurrentModificationException.
+      val propNames = props.stringPropertyNames()
+      propNames.asScala collectFirst { case k if k endsWith ".boot.class.path" => props.getProperty(k) } getOrElse ""
+    }
 
     /** Environment variables which java pays attention to so it
      *  seems we do as well.
      */
     def classPathEnv: String        =  envOrElse("CLASSPATH", "")
     def sourcePathEnv: String       =  envOrElse("SOURCEPATH", "")
 
-    def javaBootClassPath: String   = propOrElse("sun.boot.class.path", searchForBootClasspath)
+    //using propOrNone/getOrElse instead of propOrElse so that searchForBootClasspath is lazy evaluated
+    def javaBootClassPath: String   = propOrNone("sun.boot.class.path") getOrElse searchForBootClasspath
```

Comment:
this change is a second commit, that is not part of the initial one that I was porting. But I think this improvement makes sense.

Reply:
This is not an improvement. `propOrElse("user.dir", ???)`. This is the property that was removed at some point, I don't remember the details or naming. I see it is in JDK 8 but not now.

Reply:
oh, scala 2 at least supplied `scala.boot.class.path` as a workaround.

Reply:
> This is not an improvement

`propOrElse` is not lazy like in scala2, right? So this was an improvement...

Reply:
That's a shame.